catalogs/catalog-lakehouse-paimon/src/main/java/com/datastrato/gravitino/catalog/lakehouse/paimon/authentication/kerberos/FilesystemBackendProxy.java

```diff
@@ -0,0 +1,137 @@
+import org.apache.paimon.fs.hadoop.HadoopFileIO;
+import org.apache.paimon.options.CatalogOptions;
+import org.apache.paimon.options.Options;
+import org.apache.paimon.utils.Pair;
+import org.apache.paimon.utils.Preconditions;
+
+/**
+ * Proxy class for FilesystemCatalog to support kerberos authentication. We can also make
+ * FilesystemCatalog as a generic type and pass it as a parameter to the constructor.
+ */
+public class FilesystemBackendProxy implements MethodInterceptor {
+
+  private FileSystemCatalog target;
+  private final CatalogContext catalogContext;
+  private final String kerberosRealm;
+  private final UserGroupInformation proxyUser;
+
+  public FilesystemBackendProxy(
+      FileSystemCatalog target, CatalogContext catalogContext, String kerberosRealm) {
+    this.target = target;
+    this.catalogContext = catalogContext;
+    this.kerberosRealm = kerberosRealm;
+    try {
+      proxyUser = UserGroupInformation.getCurrentUser();
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to get current user", e);
+    }
+  }
+
+  @Override
+  public Object intercept(Object o, Method method, Object[] objects, MethodProxy methodProxy)
+      throws Throwable {
+
+    String proxyKerberosPrincipalName = PrincipalUtils.getCurrentPrincipal().getName();
+    if (!proxyKerberosPrincipalName.contains("@")) {
+      proxyKerberosPrincipalName =
+          String.format("%s@%s", proxyKerberosPrincipalName, kerberosRealm);
+    }
+
+    UserGroupInformation realUser =
+        UserGroupInformation.createProxyUser(proxyKerberosPrincipalName, proxyUser);
+
+    return realUser.doAs(
+        (PrivilegedExceptionAction<Object>)
+            () -> {
+              try {
+                updateFileIO(catalogContext);
+                return methodProxy.invoke(target, objects);
+              } catch (Throwable e) {
+                if (RuntimeException.class.isAssignableFrom(e.getClass())) {
+                  throw (RuntimeException) e;
+                }
+                throw new RuntimeException("Failed to invoke method", e);
+              }
+            });
+  }
+
+  public FileSystemCatalog getProxy() {
+    Enhancer e = new Enhancer();
+    e.setClassLoader(target.getClass().getClassLoader());
+    e.setSuperclass(target.getClass());
+    e.setCallback(this);
+    // FileSystemCatalog does not have a no argument constructor.
+    return (FileSystemCatalog)
+        e.create(
+            new Class[] {FileIO.class, Path.class, Options.class},
+            new Object[] {
+              target.fileIO(), new Path(target.warehouse()), new Options(target.options())
+            });
+  }
+
+  /**
+   * Before invoking the doAs method, a Filesystem client has been created when initializing the
+   * FilesystemCatalog. Therefore, the Filesystem client should be recreated within the doAs method
+   * to ensure the ugi corresponds to the impersonated user.
+   */
+  private void updateFileIO(CatalogContext catalogContext)
```

Comment:
This way seems a little tricky. We would better push Paimon community to modify here to provider more convenient interface.

Reply:
In general, the `fileIO` in the paimon catalog instance will not change, and it seems unreasonable to provide it with a new interface on the paimon side.
It seems more like a special case of gravitons, i prefer to stay the same, WDYT? cc @qqqttt123 @yuqi1129

Reply:
> In general, the `fileIO` in the paimon catalog instance will not change, and it seems unreasonable to provide it with a new interface on the paimon side. It seems more like a special case of gravitino, i prefer to stay the same, WDYT? cc @qqqttt123 @yuqi1129

Hi @FANNG1 may I ask what you think of this comment?

Reply:
it seems hacky, and if two users access Gravitino Paimon catalog at the same time, is the real user correct? And I wonder how Iceberg catalog resolve this, cc @yuqi1129

Reply:
@FANNG1 
For Iceberg with the Hive backend, different users will use different Hive clients to access HMS.

Reply:
> t seems more like a special case of gravitino, i prefer to stay the

So I was wondering, could we get the community to support this feature any time soon? Let the community support it always in the first priority, but it can't be done in a short time, I'm okay with the current changes.

Reply:
> it seems hacky, and if two users access Gravitino Paimon catalog at the same time, is the real user correct? And I wonder how Iceberg catalog resolve this, cc @yuqi1129

it might be incorrect.

Reply:
> @FANNG1 For Iceberg with the Hive backend, different users will use different Hive clients to access HMS.

maybe i can create new catalog instances for different users in the proxyPlugin and cache them, instead of using `target` object to access hdfs.
and we can also avoid using reflection. cc @yuqi1129 @FANNG1

Reply:
`create new catalog instances for different users` also seems hacky for me :(

Reply:
In the Iceberg catalog with the hive backend,  we will create a new client pool for each user and cache it for the specific user. 
I am not sure whether the Paimon catalog can also adopt this strategy.

Reply:
> `create new catalog instances for different users` also seems hacky for me :(

Hi @FANNG1 do you have any good ideas for this issue to share?

> In the Iceberg catalog with the hive backend, we will create a new client pool for each user and cache it for the specific user. I am not sure whether the Paimon catalog can also adopt this strategy.

@FANNG1 i pushed the newly code, could you help take a look?

Reply:
https://github.com/apache/gravitino/pull/3970#issuecomment-2208132573